win32.c

```diff
@@ -354,7 +354,7 @@ enet_socket_receive (ENetSocket socket,
 {
     INT sinLength = sizeof (struct sockaddr_in);
     DWORD flags = 0,
-          recvLength;
+          recvLength = 0;
```

Comment:
[Documentation for WSARecvFrom](https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsarecvfrom):

> If no error occurs and the receive operation has completed immediately, WSARecvFrom returns zero. In this case, the completion routine will have already been scheduled to be called once the calling thread is in the alertable state. Otherwise, a value of SOCKET_ERROR is returned, and a specific error code can be retrieved by calling WSAGetLastError. The error code WSA_IO_PENDING indicates that the overlapped operation has been successfully initiated and that completion will be indicated at a later time. 

This code does not use overlapped. So `WSA_IO_PENDING` should never happen. It will either return 0, or `SOCKET_ERROR`. This means that `recvLength` will only be used if the function returned immediately and without error (which is handled explicitly).

> lpNumberOfBytesRecvd: A pointer to the number of bytes received by this call if the WSARecvFrom operation completes immediately.

So this is always valid.

---

Therefore this change should not be necessary. The change would imply that `recvLength` might be used without being modified by `WSARecvFrom` which would violate the MS docs. So this makes the enet code worse in my opinion. To me it's equivalent to measuring the time before and after `Sleep()` to make sure that it actually worked - which is not necessary.

If you want to make this code safer (it is somewhat unsafe in my opinion):

- Check the return value of `WSARecvFrom` to only process the packet if the return value is 0 (and not just `!= SOCKET_ERROR`).
- Use the existing error handler for `SOCKET_ERROR`.
- Cause an error / assert for any other return value as it might be new return values added to Windows in the future.

If you got random values from this function, then I'd suggest to look for errors in your drivers (unlikely) or your application (likely). It could be something like random memory corruption.

Reply:
> If you got random values from this function, then I'd suggest to look for errors in your drivers (unlikely) or your application (likely). It could be something like random memory corruption.

You are missing the point that according to C99 _if an object that has automatic storage duration is not initialized explicitly, its value is indeterminate_ which is exactly the case, and this value is returned by the function directly. Only external and static variables are guaranteed to be initialized to zero, while local variables might contain garbage.

Reply:
If, for whatever reasons, `WSARecvFrom()` will not return `SOCKET_ERROR` (operating system has bugs too, etc), and if it doesn't fill `recvLength` internally, it's possible that this function might return a garbage value. I believe this should be covered, at least to ensure consistency.

Reply:
> You are missing the point that according to C99 if an object that has automatic storage duration is not initialized explicitly, its value is indeterminate which is exactly the case, 

It's *exactly* not the case for the reasons I explained.

I'm well aware of the behaviour in C - my entire explanation above is based on the variable containing garbage until `WSARecvFrom` sets it (which might not happen in the `SOCKET_ERROR` case). I therefore explained that whenever it's returned it will always have been set.
Initializing here implies that initialized value might be returned some time - it never should be, so it's bad practice to imply it is.

> If, for whatever reasons, `WSARecvFrom()` will not return `SOCKET_ERROR` (operating system has bugs too, etc), and if it doesn't fill `recvLength` internally, it's possible that this function might return a garbage value. 

I have explained this with the `Sleep` example.
Your code should assume that the functions work around as advertised. There's no errata for these functions so they are probably fine. If they are broken, enet *could* have a workaround (to avoid it for people without patches in their OS or while the bug is being fixed upstream). If this function is really buggy it should also be reported to MS. However, I personally strongly doubt they are broken - as also explained above.

---

By your logic we'd also have to use `if (x > 5) { if (x > 5) { [...] }} ` to make sure that the compiler doesn't have a bug either. Writing code with the assumption that underlying functions are broken is very bad practice.